### PR TITLE
Rungroup improvements: logging

### PR DESCRIPTION
**Pull request by RebeccaMahany**
Relates to https://github.com/kolide/launcher/issues/1205

Builds on oklog/rungroup to add logging, making it possible to identify which actors are not shutting down in a timely manner.

### CODE REVIEW

**Review by James-Pickett — COMMENTED** [review]
This is really good! I think it may just need some tests to be shippable.

File: pkg/rungroup/rungroup.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package rungroup
```
Comment (directionless) [inline pkg/rungroup/rungroup.go:1]: Ah. So this is a _replacement_ for oklog's rungroup, not simple a wrapper. Seems legit

File: pkg/rungroup/rungroup.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+func (g *Group) Run() error {
+	if len(g.actors) == 0 {
+		return nil
+	}
+
+	// Run each actor.
+	level.Debug(g.logger).Log("msg", "starting all actors", "actor_count", len(g.actors))
+	errors := make(chan actorError, len(g.actors))
+	for _, a := range g.actors {
+		go func(a rungroupActor) {
```
Comment (directionless) [inline pkg/rungroup/rungroup.go:53]: This is an unusual usage of an anonymous function.

If you're going to declare one that that takes an `actor` you could make a `func (g *group) start(a rungroupActor)`

or if it's anonymous, just use the `a` defined in the forloop

File: pkg/rungroup/rungroup.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			err := a.execute()
+			errors <- actorError{
+				errorSourceName: a.name,
+				err:             err,
+			}
+		}(a)
+	}
+
+	// Wait for the first actor to stop.
+	err := <-errors
```
Comment (directionless) [inline pkg/rungroup/rungroup.go:64]: Is this of type `error`? If not, calling it `err` feels confusing

File: pkg/rungroup/rungroup.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	level.Debug(g.logger).Log("msg", "received interrupt error from first actor -- shutting down other actors", "err", err)
+
+	// Signal all actors to stop.
+	for _, a := range g.actors {
+		level.Debug(g.logger).Log("msg", "interrupting actor", "actor", a.name)
+		a.interrupt(err.err)
```
Comment (directionless) [inline pkg/rungroup/rungroup.go:70]: Do we want to pass `nil` here? or leave that for future work?

File: pkg/rungroup/rungroup.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	// Wait for all actors to stop.
+	for i := 1; i < cap(errors); i++ {
+		e := <-errors
+		level.Debug(g.logger).Log("msg", "successfully interrupted actor", "actor", e.errorSourceName, "index", i)
+	}
```
Comment (directionless) [inline pkg/rungroup/rungroup.go:77]: I know that this is a copy from the original, but this is a weird idiom. Feels like wait groups would be more sensible (if harder to add logging for). Eh, maybe we leave it and adjust later?

**Review by directionless — COMMENTED** [review]
I think it's fine. I think it's pulling in a lot of weird cruft from the original, but we can iterate in subsequent steps

File: pkg/rungroup/rungroup.go (hunk 3 above)
Comment (RebeccaMahany) [reply pkg/rungroup/rungroup.go]: You're right, updated!

File: pkg/rungroup/rungroup.go (hunk 4 above)
Comment (RebeccaMahany) [reply pkg/rungroup/rungroup.go]: I don't feel strongly about it -- I'd been erring on the side of not adjusting any logic for now

**Review by James-Pickett — APPROVED** [review]

File: pkg/rungroup/rungroup.go (hunk 2 above)
Comment (directionless) [reply pkg/rungroup/rungroup.go:53]: Needing `a := a` is a good point. I dunno, it's not usually what I see in for loops.

File: pkg/rungroup/rungroup.go (hunk 4 above)
Comment (directionless) [reply pkg/rungroup/rungroup.go]: legit!

**Pull request merged**